File: Day55_SpiMat.cpp
#include <bits/stdc++.h>
using namespace std;
ifstream in("input.txt");
ofstream out("out.txt");

vector<int> spirallyTraverse(vector<vector<int>> matrix, int rr, int cc)
{
    int top = 0, bot = rr, l = 0, r = cc;
    vector<int> v;
    while (l < r and top < bot)
    {
        for (int i = l; i < r; i++)
        {
            v.push_back(matrix[top][i]);
        }
        top++;
        for (int i = top; i < bot; i++)
        {
            v.push_back(matrix[i][r - 1]);
        }
        r--;
        if (top < bot)
        {
            for (int i = r - 1; i >= l; i--)
            {
                v.push_back(matrix[bot - 1][i]);
            }
            bot--;
        }
        if (l < r)
        {
            for (int i = bot - 1; i >= top; i--)
            {
                v.push_back(matrix[i][l]);
            }
            l++;
        }
    }
    return v;
}

int main()
{
    int r, c;
    in >> r >> c;
    vector<int> v(c);
    vector<vector<int>> matrix(r, v);
    for (int i = 0; i < r; i++)
    {
        for (int j = 0; j < c; j++)
        {
            in >> matrix[i][j];
            cout << matrix[i][j] << " ";
        }
        cout << endl;
    }
    auto ans = spirallyTraverse(matrix, r, c);
    for (auto &i : ans)
    {
        cout << i << " ";
    }
    cout << endl;

    return 0;
}

File: Day55_SpiMat.py
def spirallyTraverse(matrix, r, c): 
    top = 0
    bot = r
    lef = 0
    rig = c
    l = []
    while(lef < rig and top < bot):
        for i in range(lef, rig):
            l.append(matrix[top][i])
        top += 1
        for i in range(top, bot):
            l.append(matrix[i][rig - 1])
        rig -= 1
        if(top < bot):
            for i in range(rig-1, lef-1, -1):
                l.append(matrix[bot-1][i])
            bot -= 1
        if(lef < rig):
            for i in range(bot-1, top-1, -1):
                l.append(matrix[i][lef])
            lef += 1
    return l

m = [[6, 6, 2, 28, 2], [12, 26, 3, 28, 7], [22, 25, 3, 4, 23]]
print(m)
print(spirallyTraverse(m, 3, 5))